src/mame/tandberg/tdv2100_disp_logic.cpp

```diff
@@ -0,0 +1,1063 @@
+				case 0x16:
+					if(!clear_ascii_hs_by_cleark)
+					{
+						m_write_enql_cb(0);
+						m_write_ackl_cb(0);
+						m_write_nakl_cb(0);
+					}
+					break;
+
+				default:
+					break;
+			}
+		}
+		else
+		{
+			data_to_display(byte);
+		}
+	}
+}
+
+void tandberg_tdv2100_disp_logic_device::data_to_display(uint8_t byte)
+{
+	if(m_underline_input)
+	{
+		byte = byte|0x80;
+	}
+
+	m_vram[get_ram_addr(m_cursor_row, m_cursor_col)] = byte;
+	advance_cursor();
+}
+
+void tandberg_tdv2100_disp_logic_device::advance_cursor()
+{
+	bool auto_cr_lf = BIT(m_dsw_u61->read(), 2);
+	if((m_cursor_col&0x4f) != 79 || auto_cr_lf)
+	{
+		place_cursor(m_cursor_row, m_cursor_col+1);
+	}
+
+	m_speed_check = true;
+	m_speed_ctrl->adjust(attotime::from_usec(16121));
+}
+
+void tandberg_tdv2100_disp_logic_device::place_cursor(int row, int col)
+{
+	bool in_eol_region = ((m_cursor_col&0x48) == 72);
+
+	m_cursor_col = col&0x7f;
+	m_cursor_row = row&0x1f;
+
+	if((m_cursor_col&0x48) == 72 && !in_eol_region && !m_speed_check)
+	{
+		// End of line warning beep
+		char_to_display(0x07);
+	}
+	if((m_cursor_col&0x50) == 80)
+	{
+		// End of line action
+		char_to_display(0x0d);
+		char_to_display(0x0a);
+	}
+}
+
+void tandberg_tdv2100_disp_logic_device::clear_key_handler()
+{
+	bool clear_ascii_hs_by_cleark = ((m_dsw_u39->read()&0x003) == 0x001);
+	m_write_errorl_cb(0);
+	if(clear_ascii_hs_by_cleark)
+	{
+		m_write_enql_cb(0);
+		m_write_ackl_cb(0);
+		m_write_nakl_cb(0);
+	}
+}
+
+int tandberg_tdv2100_disp_logic_device::get_ram_addr(int row, int col)
+{
+	int abs_row = (row&0x1f) + m_page_roll;
+	return ((abs_row<<7) + (col&0x7f) + (m_addr_offsets[abs_row]<<4))&0x7ff;
+}
+
+void tandberg_tdv2100_disp_logic_device::erase_row(int row)
+{
+	for(int col = 0; col<80; col++)
+	{
+		m_vram[get_ram_addr(row, col)] = 0;
+	}
+}
+
+TIMER_CALLBACK_MEMBER(tandberg_tdv2100_disp_logic_device::end_beep)
+{
+	m_beep->set_state(0);
+}
+
+TIMER_CALLBACK_MEMBER(tandberg_tdv2100_disp_logic_device::expire_speed_check)
+{
+	m_speed_check = false;
+}
+
+void tandberg_tdv2100_disp_logic_device::vblank(int state)
+{
+	if(state && !m_vblank_state)
+	{
+		// TODO: VSync interrupt for CPU interface
+		m_frame_counter++;
+		m_screen->update_now();
+	}
+	m_vblank_state = state;
+}
+
+int tandberg_tdv2100_disp_logic_device::get_attribute(int at_row, int at_col, int attribute, int from_row, int from_col)
+{
+	for(int char_row = from_row; char_row <= at_row; char_row++)
+	{
+		for(int char_col = (char_row == from_row)? from_col : 0; char_col < 80; char_col++)
+		{
+			int const vram_data = m_vram[get_ram_addr(char_row, char_col)];
+			if(vram_data&0x80)
+			{
+				attribute = (vram_data&0x70)>>4;
+			}
+			if(char_row == at_row && char_col == at_col)
+			{
+				break;
+			}
+		}
+	}
+	return attribute;
+}
+
+uint32_t tandberg_tdv2100_disp_logic_device::screen_update(screen_device &screen, bitmap_rgb32 &bitmap, const rectangle &cliprect)
+{
```

Comment:
For what it’s worth, because of the way this function always draws entire characters, it won’t work properly if something triggers a partial update on a screen line that isn’t the first row of a character.  The previous rows of the first line of characters intersecting the clipping rectangle will get redrawn.

If you *do* want to cater for cases like this, do you know how the display logic deals with attributes for each row of a line of characters?  For example does it latch the attributes when it starts drawing the first row of a line of characters, then reuse that value when it starts each subsequent row of the line?  Or does it have an line-sized attribute buffer it fills on the first row, or does it use some other approach?

Reply:
The attribute is latched anywhere through the scan, and reset only on VSync (not per line). Any character on screen can store an attribute change, but with default DIP switches these will not be drawn.

So I have already accounted for this since each time screen_update is called, I scan from upper left (0,0) to get the attribute for the first character in the rectangle section before any drawing is done. The attribute variable is in other words not shared between individual calls to screen_update. In this way, overlapping draws will not mess up the attribute.

Reply:
You’re missing the point.  You’re drawing all the rows of pixels in a character in one go before moving to the next character horizontally.

It isn’t possible for the hardware to do this since the raster display scans an entire line of pixels before moving to the next line.  After drawing the first row of pixels in a character, it draws the first row of pixels in the next character to the right, and so on.  Then on the next line it draws the second row of pixels for each character in the same line.

For this to work, it needs to be able to restore the attribute state for the beginning of the line of characters each time it starts drawing a line of pixels that isn’t correspond to the first row of pixels in a line of characters.  Usually this is done either by latching the attribute state at the beginning of screen line corresponding to the first row of pixels in a line of characters, then restoring that state at the start of subsequent rows of pixels for the same line of characters.  However, other approaches are also possible, such as keeping a line width attribute buffer and filling it while drawing the first row of pixels in a line of characters.

> So I have already accounted for this since each time screen_update is called, I scan from upper left (0,0) to get the attribute for the first character in the rectangle section before any drawing is done.

Think about what happens when video RAM or registers change part way through a frame.  This is what causes visual artefacts like “tearing”.  The portion of the raster that has already been drawn will be unaffected, but the part that has yet to be drawn will use the new values (assuming the RAM/registers aren’t double-buffered).

In MAME, we deal with this using partial updates.  When something that affects video display changes, you trigger a partial update before setting the values so the parts of the screen that would be drawn before the change takes effect won’t see the change.

Consider what would happen on a partial update with your code as written now:
* The code draws all rows of any line of characters that intersects the clipping rectangle.  Supposed a partial update is triggered after three rows of a line of characters has been drawn.  With your current implementation, those three rows will be redrawn when the remaining rows of that line of characters are drawn.
* On every update, you scan attributes from the start of video RAM.  If attribute characters are changed in the part of the frame that has already been drawn in a partial update, they will still be applied in subsequent updates for the same frame.

Reply:
> For this to work, it needs to be able to restore the attribute state for the beginning of the line of characters each time it starts drawing a line of pixels that doesn’t correspond to the first row of pixels in a line of characters. Usually this is done either by latching the attribute state at the beginning of screen line corresponding to the first row of pixels in a line of characters, then restoring that state at the start of subsequent rows of pixels for the same line of characters.

Sorry for not going in too much detail, but this is exactly what the hardware does. The attribute is latched in a backup register at the end of the last line of each row, and recovered at the start of each line in the next row. See "Attribute Memory/Latch" on page 33/35:
https://www.technikum29.de/de/geraete/Siemens_6610/Siemens_6610_PDFs/Siemens_6610_Wartungskurs_Reg_I_Kap_0_Display_Logic_I_und_II.pdf

> Think about what happens when video RAM or registers change part way through a frame. This is what causes visual artefacts like “tearing”. The portion of the raster that has already been drawn will be unaffected, but the part that has yet to be drawn will use the new values (assuming the RAM/registers aren’t double-buffered).

The hardware double-buffers display RAM one row at a time, and halts any other access when the buffer is swapped out (also happens during last line of a row of characters). See "Refresh Memory", page 21/23 in the pdf linked to.

> On every update, you scan attributes from the start of video RAM. If attribute characters are changed in the part of the frame that has already been drawn in a partial update, they will still be applied in subsequent updates for the same frame.

The only way around this is to force a full redraw every time an attribute is changed.

Reply:
Ok, I have made it so that it only scans attribute between the previous draw rectangle and the start of the current draw rectangle.